### ISSUE TITLE
How to fix nan problem ?

### ISSUE THREAD

**Issue by liesun1994** [clab/dynet#990]
When I am trying to train a NMT model based on dynet. Errors always occured after training multi-batches.
My code is listed as follows: 
`soft_val = softmax(score);`
`loss = pickneglogsoftmax(soft_val, vals)`
where vals is target indices.
And after training my model a while , errors occured . 
`terminate called after throwing an instance of 'std::runtime_error'
  what():  Magnitude of gradient is bad: -nan`
Something wrong with my code ?

**Comment by neubig** [comment]
The most common cause of "nan" is when you put an input into a function that is out of the range of the function, for example dividing by zero, taking the log of zero or a negative number, or taking an exponent of a very large number. In these cases, nan will often occur suddenly and without warning.

It also sometimes happen because your learning rate is too high and gradients explode. In this case numbers will expand to become larger and larger until you get a nan.

nans can be debugged using immediate computation and value checking described here: http://dynet.readthedocs.io/en/latest/debugging.html

**Comment by pmichel31415** [comment]
pickneglogsoftmax is (as mentionned in the documentation) already performing a softmax. So you should apply it to the scores directly and not softmax(score)

**Comment by liesun1994** [comment]
Thanks for your reply . I have to check my code and trying to figure it out .

**Comment by liesun1994** [comment]
@pmichel31415 I am wondering wheather softmax function is able to handle nearly zero problems and minus max number .

**Comment by chunyang-wen** [comment]
you can use:

```python
dynet.renew_cg(immediate_compute = True, check_validity = True)
```

It can be used to debug and find bad values as soon as possible.

**Comment by liesun1994** [comment]
@chunyang-wen Yeap,  I am using c++ and the command is
 `ComputationGraph cg;`
`cg.set_immediate_compute(true);`
`cg.set_check_validity(true);`
I used in my code and error occured too. 
`is_valid() is not implement on GPU.`
aha.
I am printing each value such as `softmax`，`-log(picked_soft)` step by step . 
Hope it can be fixed .

**Comment by chunyang-wen** [comment]
@liesun1994 You can turn off GPU first.

**Comment by liesun1994** [comment]
@chunyang-wen  That might be quite slow . And I have no idea how to turn off GPU ?
